stoppable_workers.go

```diff
@@ -0,0 +1,64 @@
+package utils
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// ErrStoppableWorkersAlreadyStopped is returned by Add when the StoppableWorkers
+// instance has already been stopped.
+var ErrStoppableWorkersAlreadyStopped = errors.New("cannot add worker: already stopped")
+
+// StoppableWorkers is a collection of goroutines that can be stopped at a
+// later time.
+type StoppableWorkers struct {
+	mu         sync.RWMutex
+	ctx        context.Context
+	cancelFunc func()
+
+	workers sync.WaitGroup
+}
+
+// NewStoppableWorkers creates a new StoppableWorkers instance. The instance's
+// context will be derived from passed in context.
+func NewStoppableWorkers(ctx context.Context) *StoppableWorkers {
```

Comment:
I like the pattern of having each stoppable worker make its own context; it made sense since the components with multiple workers could all form their contexts and be self-contained in construction and close. Want to make another creator function that gives you your behaviour instead?

Reply:
Yep can do.